0023-communication-protocol.adoc

```diff
@@ -0,0 +1,383 @@
+
+[source,json]
+----
+{
+  "swap_id": "713b0c93-53a1-46b2-9418-de8d46750b79", // <1>
+  "address": "0xd3935B8cCA773146Cda591eF468Ce141BD6d57f4" // <2>
+}
+----
+1. The swap-id from the `/comit/swap/announce/1.0.0/` protocol
+2. An Ethereum address
+
+Protocol identifier: `/comit/swap/identity/lightning/1.0.0`
+
+[source,json]
+----
+{
+  "swap_id": "713b0c93-53a1-46b2-9418-de8d46750b79", // <1>
+  "pubkey": "0236055ab508ec9495c07a3968f7ee9a1b5983f70e24d362f677d2619419601052" // <2>
+}
+----
+1. The swap-id from the `/comit/swap/announce/1.0.0/` protocol
+2. A Lightning node public key (taken from `lncli getinfo`)
+
+==== Why do we not need more protocols?
+
+Given the context of a swap, all of these protocols carry unambiguous information.
+Consider the following example:
+
+Assume `Bitcoin` is the alpha-ledger and `Ethereum` is the beta-ledger.
+In said constellation, Alice's refund identity is a Bitcoin public key and here redeem identity is an Ethereum address.
+
+If Alice would send a message using the `/comit/swap/identity/bitcoin/1.0.0` protocol, it is unambiguous for Bob that Alice is sending her refund identity.
+
+More generally speaking:
+Given knowledge of the alpha and beta ledger and the role, a message can be unambiguously identified as the redeem or refund identity of an actor depending on whether the message is incoming or outgoing.
+
+This allows us to slim down the protocols to the bare minimum of sending over an identity.
+
+=== Finalizing a swap
+
+One of the downsides of splitting the communication for a single swap up into several protocols is the fact that we don't have a natural "end" of the communication.
+If all of the communication were to take place within a single protocol, closing the substream could act as "I am done communicating".
+
+Following a multiple protocol approach, the answer to how we communicate this concern is another protocol.
+
+Introducing: `/comit/swap/finalize/1.0.0`
+
+[source,json]
+----
+{
+  "swap_id": "713b0c93-53a1-46b2-9418-de8d46750b79"
+}
+----
+
+Both parties are expected to use this protocol after they have sent all the data they consider necessary for starting the execution of the swap.
+Once each party sent a finalize message AND received one, implementations should consider the swap finalized and start with the execution phase (i.e. watch for transactions, emit actions, etc)
+
+== Discussion points
+
+This section includes discussion topics that remain to be answered.
+
+=== Should we use multihash for the swap digest
+
+<<Announce-message,Section 3.2>> introduces the concept of a swap digest.
+In order to be forwards compatible with different hash algorithms, it proposes to encode the root hash as a multihash.
+That way, the hash is self-describing and the receiving party knows, which hash algorithm needs to be used to compute the digest.
+
+Do we need the flexibility of changing the hash function that is used to compute the swap digest without a breaking change to the protocol?
```

Comment:
I understand that the swap digest is independent from the remaining messages and is used for the following things: 
* it is needed to uniquely announce a swap (provided the function used to compute the digest is collision free)
* Allow to verify to known information against the swap digest, 
** detect if data is invalid 
** detect if data is missing

That being said, there might be even better data structures then a multi-hash.

If the swap digest is independent of the other messages we can _play around_ with it and see what works well for us?

Reply:
> uniquely announce a swap (provided the function used to compute the digest is collision free)

The function is unlikely to be totally collision free because it is oriented around the "offer" that both parties gave to cnd (offer ~ swap - identities).
If we include the cnd identity of the other party within the digest, the only collisions that can happen are between those two parties. If that happens, it is unlikely to be a collision but actually an announcement of a second swap (with the same parameters).

> allow to verify known information against the swap digest:

Yes. The main purpose is to verify information without having to send a lot of data.

> That being said, there might be even better data structures then a multi-hash.

Note that the question you commented on specifically questions the use of the multihash. The alternative would be an untagged hash (where we hardcode the hash function).

> If the swap digest is independent of the other messages we can play around with it and see what works well for us?

Yes that is the idea of splitting up the messages. We should be able to change one part without affecting the others.